Modules/Filtering/ImageGradient/include/itkVectorGradientMagnitudeImageFilter.h

```diff
@@ -225,17 +227,24 @@ class ITK_TEMPLATE_EXPORT VectorGradientMagnitudeImageFilter : public ImageToIma
   }
 #endif
 
-  using WeightsType = FixedArray<TRealType, VectorDimension>;
+  using ComponentWeightsType = FixedArray<TRealType, VectorDimension>;
+  using DerivativeWeightsType = FixedArray<TRealType, ImageDimension>;
+#if !defined(ITK_LEGACY_REMOVE)
+#  if __cplusplus >= 201402L
+  [[deprecated("Use DerivativeWeightsType or ComponentWeightsType instead.")]]
+#  endif
+  using WeightsType = ComponentWeightsType;
```

Comment:
Remove the `#if __cplusplus` branch. It will be handled in future PR with a macro.

Reply:
Or, if you wish, this is valid ITK code:
```cpp
#if !defined(ITK_LEGACY_REMOVE)
  using WeightsType ITK_DEPRECATED_MSG("Use DerivativeWeightsType or ComponentWeightsType instead.") =
    ComponentWeightsType;
#endif
```

Reply:
Interesting, looks like here
```
#    if (__GNUC__ * 100 + __GNUC_MINOR__) >= 409 && __cplusplus > 201103L
#      define ITK_COMPILER_CXX_ATTRIBUTE_DEPRECATED 1
#    else
#      define ITK_COMPILER_CXX_ATTRIBUTE_DEPRECATED 0
#    endif
```
is assumed C++11, but i that `[[deprecated]]` thing is definitely C++14. Probably this part

```
...
#    if defined(ITK_COMPILER_CXX_ATTRIBUTE_DEPRECATED) && ITK_COMPILER_CXX_ATTRIBUTE_DEPRECATED
#      define ITK_DEPRECATED [[deprecated]]
#      define ITK_DEPRECATED_MSG(MSG) [[deprecated(MSG)]]
...
```
will cause problems sometimes. May be `__cplusplus > 201103L` above is a bug.

About compiler intrinsic things `__attribute__ ` for GCC and `__declspec` for Visual Studio, not sure, but i use them for address alignment and it was not easy to get it working, not sure how  it works with deprecated, but for alignment one requires prepend and another append.
This things are difficult to test, sometimes they work, sometimes not. BTW, many Visual Studio version have issues with value of ___cplusplus_, updated VS 2019 don't, but earlier versions reported 1998xx and ignored cmake.

Reply:
Thanks both of you for the interest, time, and effort to propose a proper solution to the backwards compatibility issue. I've push forced. Let's see what the builds say.

Reply:
This is what i mean with prepend/append

```
#ifdef __GNUC__
#define DEPRECATED(func) func __attribute__ ((deprecated))
#elif defined(_MSC_VER)
#define DEPRECATED(func) __declspec(deprecated) func
...
```

Reply:
CI builds seem not to hiccup till now/the code is compiling,  but it might with the larger set of compilers in the dashboard.

If I understand correctly, [the above](https://github.com/InsightSoftwareConsortium/ITK/pull/2279#discussion_r571468667) should go into [CMake](https://gitlab.kitware.com/cmake/cmake/-/blob/master/Modules/WriteCompilerDetectionHeader.cmake)? A patch set might be desirable there?

Reply:
@jhlegarreta i am not sure about syntax with `using` and don't know - does it cause build error or simply doesn't work sometimes, so actually i am not ready for PR, if somebody is sure, may be. It is complicated stuff, many combinations, IMHO.

Reply:
So let's keep this in mind if at some point we witness related problems, and get this PR merged if builds are passing so as to keep moving forward. Thanks for all the useful thoughts !

Reply:
@issakomi it is tricky indeed with the different placement of the attributes among compilers :(
Maybe is worth to have a look at the third party project for these kind of `ifdefs`, as a reference: https://github.com/nemequ/hedley/blob/master/hedley.h 

For the case of deprecated functions, we already have the `itkLegacyMacro`: https://github.com/InsightSoftwareConsortium/ITK/blob/3fd1ff784029f226e976c016b06068a9af0b5d4a/Modules/Core/Common/include/itkMacro.h#L601-L618 that handles the different placement.

It seems the placement is shared among compilers with `using` at least.

Reply:
I have tested with GCC 8.3.0 and it works good with both C++11 and C++14. Also it seems to work with this GCC version both with prepend and append of __attribute__((deprecated)). Not sure about all compilers, specially older, but seems to be fine for now. Thank you.